discovery/hetzner/hcloud.go

```diff
@@ -0,0 +1,122 @@
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hetznerHcloudLabelPrefix                        = hetznerLabelPrefix + "hcloud_"
+	hetznerLabelHcloudImageName                     = hetznerHcloudLabelPrefix + "image_name"
+	hetznerLabelHcloudImageDescription              = hetznerHcloudLabelPrefix + "image_description"
+	hetznerLabelHcloudPrivateIPv4                   = hetznerHcloudLabelPrefix + "private_ipv4"
+	hetznerLabelHcloudDatacenterLocation            = hetznerHcloudLabelPrefix + "datacenter_location"
+	hetznerLabelHcloudDatacenterLocationNetworkZone = hetznerHcloudLabelPrefix + "datacenter_location_network_zone"
+	hetznerLabelHcloudCPUCores                      = hetznerHcloudLabelPrefix + "cpu_cores"
+	hetznerLabelHcloudCPUType                       = hetznerHcloudLabelPrefix + "cpu_type"
+	hetznerLabelHcloudMemoryGB                      = hetznerHcloudLabelPrefix + "memory_size_gb"
+	hetznerLabelHcloudDiskGB                        = hetznerHcloudLabelPrefix + "disk_size_gb"
+	hetznerLabelHcloudType                          = hetznerHcloudLabelPrefix + "server_type"
+)
+
+// Discovery periodically performs Hetzner Cloud requests. It implements
+// the Discoverer interface.
+type hcloudDiscovery struct {
+	*refresh.Discovery
+	client *hcloud.Client
+	port   int
+}
+
+// newHcloudDiscovery returns a new hcloudDiscovery which periodically refreshes its targets.
+func newHcloudDiscovery(conf *SDConfig, logger log.Logger) (*hcloudDiscovery, error) {
+	d := &hcloudDiscovery{
+		port: conf.Port,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false)
+	if err != nil {
+		return nil, err
+	}
+	d.client = hcloud.NewClient(
+		hcloud.WithApplication("Prometheus", version.Version),
+		hcloud.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		hcloud.WithEndpoint(conf.hcloudEndpoint),
+	)
+	return d, nil
+}
+func (d *hcloudDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	servers, err := d.client.Server.All(ctx)
+	if err != nil {
+		return nil, err
+	}
+	networks, err := d.client.Network.All(ctx)
+	if err != nil {
+		return nil, err
+	}
+	targets := make([]model.LabelSet, len(servers))
+	for i, server := range servers {
+		labels := model.LabelSet{
+			hetznerLabelPlatform:     model.LabelValue(hetznerPlatformHcloud),
+			hetznerLabelServerID:     model.LabelValue(fmt.Sprintf("%d", server.ID)),
+			hetznerLabelServerName:   model.LabelValue(server.Name),
+			hetznerLabelDatacenter:   model.LabelValue(server.Datacenter.Name),
+			hetznerLabelPublicIPv4:   model.LabelValue(server.PublicNet.IPv4.IP.String()),
+			hetznerLabelPublicIPv6:   model.LabelValue(server.PublicNet.IPv6.Network.String()),
```

Comment:
Do you mean network or IP?

Reply:
Every server (dedicated and cloud) has a /64 IPv6 Net and we never know which IP within this network is correctly configured

Reply:
if the api returns both let's have both net and ip as label

Reply:
The API does not return an IP, it always return the IPv6 Net. I will rename the Label to "public_ipv6_net"

Reply:
https://godoc.org/github.com/hetznercloud/hcloud-go/hcloud#ServerPublicNetIPv6 has ip which confused me

Reply:
Yeah on cloud servers we assume (or at least within the base images it is) that the first IP within the net is configured everytime.